prime-router/src/main/kotlin/azure/ValidateFunction.kt

```diff
@@ -0,0 +1,328 @@
+                // Only process the report if we are not checking for connection or validation.
+                if (options != Options.CheckConnections && options != Options.ValidatePayload) {
+                    val receiver = ValidationReceiver(workflowEngine, actionHistory)
+
+                    // send report on its way, either via the COVID pipeline or the full ELR pipeline
+                    receiver.validateAndMoveToProcessing(
+                        sender,
+                        validatedRequest.content,
+                        validatedRequest.defaults,
+                        options,
+                        validatedRequest.routeTo,
+                        isAsync,
+                        allowDuplicates,
+                        rawBody,
+                        payloadName
+                    )
+
+                    // return CREATED status, report submission was successful
+                    HttpStatus.CREATED
+                } else HttpStatus.OK
+            } catch (e: ActionError) {
+                actionHistory.trackLogs(e.details)
+                HttpStatus.BAD_REQUEST
+            } catch (e: IllegalArgumentException) {
+                actionHistory.trackLogs(
+                    ActionLog(InvalidReportMessage(e.message ?: "Invalid request."), type = ActionLogLevel.error)
+                )
+                HttpStatus.BAD_REQUEST
+            } catch (e: IllegalStateException) {
+                actionHistory.trackLogs(
+                    ActionLog(InvalidReportMessage(e.message ?: "Invalid request."), type = ActionLogLevel.error)
+                )
+                HttpStatus.BAD_REQUEST
+            }
+
+        val submission = DetailedSubmissionHistory(
+            1,
+            TaskAction.none,
+            OffsetDateTime.now(),
+            httpStatus.value(),
+            mutableListOf<DetailedReport>(),
+            actionHistory.actionLogs.map {
+                DetailedActionLog(
+                    it.scope,
+                    it.reportId,
+                    it.index,
+                    it.trackingId,
+                    it.type,
+                    it.detail
+                )
+            }
+        )
+
+        // set status for validation response
+        submission.overallStatus = if (httpStatus == HttpStatus.BAD_REQUEST)
+            DetailedSubmissionHistory.Status.ERROR
+        else
+            DetailedSubmissionHistory.Status.VALID
+
+        return request.createResponseBuilder(httpStatus)
+            .header(HttpHeaders.CONTENT_TYPE, "application/json")
+            .body(
+                JacksonMapperUtilities.allowUnknownsMapper
+                    .writeValueAsString(submission)
+            )
+            .build()
+    }
+
+    /**
+     * Extract client header from request headers or query string parameters
+     * @param request the http request message from the client
+     */
+    private fun extractClient(request: HttpRequestMessage<String?>): String {
```

Comment:
This function is an identical duplicate with the one in ReportFunction. Perhaps full out into a shared base class?